libnmstate/nm/checkpoint.py

```diff
@@ -92,21 +49,97 @@ def __exit__(self, exc_type, exc_val, exc_tb):
-            CHECKPOINT_CREATE_FLAG_DESTROY_ALL |
-            CHECKPOINT_CREATE_FLAG_DELETE_NEW_CONNECTIONS |
-            CHECKPOINT_CREATE_FLAG_DISCONNECT_NEW_DEVICES
-        )
-        cp = self._manager.interface.CheckpointCreate(devs, timeout, cp_flags)
-        logging.debug('Checkpoint %s created for all devices: %s', cp, timeout)
-        self._checkpoint = cp
+        cp_flags = (nmclient.NM.CheckpointCreateFlags.DESTROY_ALL |
+                    nmclient.NM.CheckpointCreateFlags.DELETE_NEW_CONNECTIONS |
+                    nmclient.NM.CheckpointCreateFlags.DISCONNECT_NEW_DEVICES)
+        userdata = self
+
+        self._mainloop.push_action(
+            self._client.checkpoint_create,
+            devs, timeout, cp_flags, None, _checkpoint_create_callback,
+            userdata)
+
+        self._mainloop.run(timeout)
 
     def destroy(self):
-        self._manager.interface.CheckpointDestroy(self._checkpoint)
-        logging.debug('Checkpoint %s destroyed', self._checkpoint)
+        if self._checkpoint:
+            userdata = self
+            self._mainloop.push_action(
+                self._client.checkpoint_destroy,
+                self._checkpoint.get_path(), None,
+                _checkpoint_destroy_callback, userdata)
+            self._mainloop.run(self._timeout)
+        else:
+            logging.warning('No checkpoint to destroy')
 
     def rollback(self):
-        result = self._manager.interface.CheckpointRollback(self._checkpoint)
-        logging.debug('Checkpoint %s rollback executed: %s',
-                      self._checkpoint, result)
-        return result
+        if self._checkpoint:
+            userdata = self
+            self._mainloop.push_action(
+                self._client.checkpoint_rollback,
+                self._checkpoint.get_path(), None,
+                _checkpoint_rollback_callback, userdata)
+            self._mainloop.run(self._timeout)
+        else:
+            logging.warning('No checkpoint to rollback')
+
+def _checkpoint_create_callback(client, result, data):
+    mainloop = data._mainloop
+    try:
+        cp = client.checkpoint_create_finish(result)
+        data._checkpoint = cp
+        if cp is None:
+            mainloop.quit('Checkpoint creating failed: error=unknown')
+        else:
+            logging.debug('Checkpoint %s created for all devices',
+                          cp.get_path())
+            mainloop.execute_next_action()
+    except Exception as e:
+        if mainloop.is_action_canceled(e):
+            logging.debug('Checkpoint creating canceled: error=%s', e)
+        else:
+            mainloop.quit('Checkpoint creating failed: error={}'.format(e))
+        return
+
+
+def _checkpoint_rollback_callback(client, result, data):
+    mainloop = data._mainloop
+    cp = data._checkpoint
+    try:
+        ret = client.checkpoint_rollback_finish(result)
+        logging.debug('Checkpoint %s rollback executed', cp.get_path())
+        for path in ret:
```

Comment:
Please extract this for/block to a func.

Reply:
Done.